server/src/services/library.service.ts

```diff
@@ -598,106 +622,96 @@ export class LibraryService extends EventEmitter {
+    const pathValidation = await Promise.all(
+      library.importPaths.map(async (importPath) => await this.validateImportPath(importPath)),
     );
 
-    this.logger.verbose(`Crawled asset paths paginated`);
-
-    const shouldScanAll = job.refreshAllFiles || job.refreshModifiedFiles;
-    for await (const page of pagination) {
-      for (const asset of page) {
-        const isOffline = !crawledAssetPaths.has(asset.originalPath);
-        if (isOffline && !asset.isOffline) {
-          assetIdsToMarkOffline.push(asset.id);
-          this.logger.verbose(`Added to mark-offline list: ${asset.originalPath}`);
+    const validImportPaths = pathValidation
+      .map((validation) => {
+        if (!validation.isValid) {
+          this.logger.error(`Skipping invalid import path: ${validation.importPath}. Reason: ${validation.message}`);
         }
+        return validation;
+      })
+      .filter((validation) => validation.isValid)
+      .map((validation) => validation.importPath)
+      .map((importPath) => path.normalize(importPath));
 
-        if (!isOffline && asset.isOffline) {
-          assetIdsToMarkOnline.push(asset.id);
-          this.logger.verbose(`Added to mark-online list: ${asset.originalPath}`);
-        }
+    const crawledAssets = this.storageRepository.walk({
+      pathsToCrawl: validImportPaths,
+      exclusionPatterns: library.exclusionPatterns,
+    });
 
-        if (!shouldScanAll) {
-          crawledAssetPaths.delete(asset.originalPath);
-        }
+    const onlineAssets = usePagination(LIBRARY_SCAN_BATCH_SIZE, (pagination) =>
+      this.assetRepository.getWith(pagination, WithProperty.IS_ONLINE, job.id),
+    );
+
+    let crawlDone = false;
+    let existingAssetsDone = false;
+    let crawlCounter = 0;
+    let existingAssetCounter = 0;
+
+    const checkIfOnlineAssetsAreOffline = async () => {
+      const existingAssetPage = await onlineAssets.next();
+      existingAssetsDone = existingAssetPage.done ?? true;
+
+      if (existingAssetPage.value) {
+        existingAssetCounter += existingAssetPage.value.length;
+        this.logger.log(
+          `Queuing online check of ${existingAssetPage.value.length} asset(s) in library ${library.id}...`,
+        );
+        await this.jobRepository.queueAll(
+          existingAssetPage.value.map((asset: AssetEntity) => ({
+            name: JobName.LIBRARY_CHECK_OFFLINE,
+            data: { id: asset.id, importPaths: validImportPaths },
+          })),
+        );
       }
-    }
+    };
 
-    this.logger.verbose(`Crawled assets have been checked for online/offline status`);
+    let crawledAssetPaths: string[] = [];
 
-    if (assetIdsToMarkOffline.length > 0) {
-      this.logger.debug(`Found ${assetIdsToMarkOffline.length} offline asset(s) previously marked as online`);
-      await this.assetRepository.updateAll(assetIdsToMarkOffline, { isOffline: true });
-    }
+    while (!crawlDone) {
```

Comment:
This is a bit weird. We should be able to use a normal (async) for loop here instead.

Reply:
I don't understand. Why can't you do
```ts
for await (const crawlResult of crawledAssets) {
...
}
```
?

Reply:
The code needs to do two tasks, and to make the system feel more responsive I don't want to wait for the first (potentially very long) job to be worked on before the next jobs start. We therefore interleave the two jobs in between

Reply:
This reply does not make any sense to me. Doesn't this semantically do the exact same thing?

Reply:
Done